src/Servers/Kestrel/Transport.Quic/src/Internal/QuicStreamContext.cs

```diff
@@ -324,7 +324,7 @@ private async ValueTask ShutdownWrite(Exception? shutdownReason)
             {
                 lock (_shutdownLock)
                 {
-                    _shutdownReason = shutdownReason ?? new ConnectionAbortedException("The Quic transport's send loop completed gracefully.");
+                    _shutdownReason = shutdownReason;
```

Comment:
This changes more than jus logging. _shutdownReason is passed to Input.Complete() in DoReceive() before falling back to any exception that might have been thrown out of the read loop.

We do the same thing in SocketConnection.cs because closing a socket usually causes the read loop to throw with unhelpful exceptions, when the reader from the stream really wants to know the root cause which is that the server closed the socket because it was done writing (in the case where shutdownReason is null). I'm not 100% sure the same logic should apply to QuicStreamContext, but it feels like the same principles apply.

Reply:
To clean up the logs, we can do what we do in SocketConnection and just include the exception message (not the exception itself) in the logs.

https://github.com/dotnet/aspnetcore/blob/master/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L350

Or maybe we could change both to pass the entire exception to the logger only when a non-null shutdownReason is supplied.

Reply:
Is the _shutdownReason exception passed to `Input.Complete(ex)` only used if `Input.Complete()` hasn't already been called somewhere else?

Right now we're allocating an exception per-request which isn't desirable. We would only want to create the exception if it is going to be used.

Reply:
Correct. It's only used if Input.Complete() hasn't already been called.

I'm be fine with setting flag in DoReceive() after calling Input.Complete() and then skipping the ConnectionAbortedException allocation if that flag is already set. We just have to be careful to do it in a way that doesn't change the StreamShutdownWrite log.